### PR TITLE
[thci] add OpenThread_BR THCI for OTBR CLI

### PR DESCRIPTION
**Pull request by simonlingoogle**
- [x] Implement THCI for [ot-br-posix](https://github.com/openthread/ot-br-posix/)
- [x] Retrieve commissioning log on OTBR
- [x] 1.1 TestHarness regression for OTBR
  - Got same regression result for old and new `OpenThread.py` on OT CLI devices
  - Got a number of cases failed for OTBR CLI (will investigate later)

**Some implementation details:**
- The new `OpenThreadTHCI` class serves as the base class for both `OpenThread` and `OpenThread_BR`. It also includes the logic of all THCI APIs. 
  * It reuses most code of the original `OpenThread` class. 
- `OpenThreadTHCI` defines a number of abstract methods that is to be defined by derived classes (i.e. `OpenThread` and `OpenThread_BR`)
  * **_connect**: connect to the underlying device
  * **_disconnect**: disconnect from the underlying device
  * **_cliReadLine**: read the next line from OT CLI, or None if no data
  * **_cliWriteLine**: write a line of command to the OT CLI
  * **_onCommissionStart**: notify commissioning starts
  * **_onCommissionStop**: notify commissioning stops

**Note that `OpenThreadTHCI` assumes that the underlying device must provide the same OT CLI interface**, and all logic for interacting with OT CLI are implemented in `OpenThreadTHCI` by calling `_cliReadLine` and `_cliWriteLine`.

### CODE REVIEW

File: tools/harness-thci/OpenThread_Uno.py [hunk 1]
```diff
@@ -0,0 +1,272 @@
+#!/usr/bin/env python
```
Comment (wgtdkp) [inline tools/harness-thci/OpenThread_Uno.py:1]: `OpenThread_Uno.py`, `uno` is the internal code code, should we use the name for public?

File: tools/harness-thci/OpenThread.py [hunk 2]
```diff
@@ -48,218 +56,139 @@
...
-                    )
-                    if retry_times == 0:
-                        raise
-                else:
-                    break
+        if self.firmwarePrefix in self.UIStatusMsg:
+            self.deviceConnected = True
+        else:
+            self.UIStatusMsg = ('Firmware Not Matching Expecting ' +
+                                self.firmwarePrefix + ' Now is ' +
```
Comment (Irving-cl) [inline tools/harness-thci/OpenThread.py:174]:
```suggestion
                                self.firmwarePrefix + ', Now is ' +
```
Add a comma here?

File: tools/harness-thci/OpenThread.py [hunk 3]
```diff
@@ -2811,3 +2771,182 @@ def __lstrip0x(s):
...
+    def __del__(self):
+        """close the serial port connection"""
+        try:
+            self.closeConnection()
+            self.deviceConnected = False
+        except Exception as e:
+            ModuleHelper.WriteIntoDebugLogger('delete() Error: ' + str(e))
+
+    def _expect(self, expected, times=50):
+        """Find the `expected` line within `times` trials.
```
Comment (Irving-cl) [inline tools/harness-thci/OpenThread.py:2807]:
```suggestion
        """Find the `expected` line within `times` tries.
```

File: tools/harness-thci/OpenThread.py [hunk 4]
```diff
@@ -2811,3 +2771,182 @@ def __lstrip0x(s):
...
+
+        Args:
+            expected    str: the expected string
+            times       int: number of trials
```
Comment (Irving-cl) [inline tools/harness-thci/OpenThread.py:2811]:
```suggestion
            times       int: number of tries
```

File: tools/harness-thci/OpenThread.py [hunk 5]
```diff
@@ -2811,3 +2771,182 @@ def __lstrip0x(s):
...
+        if self._is_net:
+            self.handle.sendall(data)
+        else:
+            self.handle.write(data)
+
+    def _readline(self):
+        """Read exactly one line from the device
+
+        Returns:
+            None on no data
```
Comment (Irving-cl) [inline tools/harness-thci/OpenThread.py:2850]:
```suggestion
            None if no data
```

File: tools/harness-thci/OpenThread.py [hunk 6]
```diff
@@ -2811,3 +2771,182 @@ def __lstrip0x(s):
...
+        self.sleep(0.1)
+
+    def _sendCommand(self, cmd, expectEcho=True):
+        self.log("_sendCommand: %s", cmd)
+        self._sendline(cmd)
+        if expectEcho:
+            self._expect(cmd)
+
+    def _executeCommand(self, cmd):
+        """send specific command to reference unit over serial port
```
Comment (Irving-cl) [inline tools/harness-thci/OpenThread.py:2897]: On line 107 it is described as:
```
    def _executeCommand(self, cmd):
        """
        Execute the command on the device.
```
Shall we make the comment to be consistent?

File: tools/harness-thci/OpenThread_Uno.py [hunk 7]
```diff
@@ -0,0 +1,272 @@
...
+#
+# Copyright (c) 2018, The OpenThread Authors.
```
Comment (Irving-cl) [inline tools/harness-thci/OpenThread_Uno.py:3]:
```suggestion
# Copyright (c) 2020, The OpenThread Authors.
```

**Pull request merged**